# Assessment Review

The topics covered will be:

1. SQL
2. Bayesian Statistics
3. Normal Distributions
4. Statistical Tests

## Part 1: SQL

In [1]:
# Need to import pandas and sqlite3

import sqlite3
import pandas as pd

In [2]:
conn = sqlite3.connect('data/Chinook_Sqlite.sqlite')

In [3]:
pd.read_sql(
    """
    SELECT *
    FROM Customer
    LIMIT 3
    """, conn
)

CustomerId FirstName   LastName  \
0           1      Luís  Gonçalves   
1           2    Leonie     Köhler   
2           3  François   Tremblay   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                              None   
2                                              None   

                           Address                 City State  Country  \
0  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP   Brazil   
1          Theodor-Heuss-Straße 34            Stuttgart  None  Germany   
2                1498 rue Bélanger             Montréal    QC   Canada   

  PostalCode               Phone                 Fax                  Email  \
0  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   luisg@embraer.com.br   
1      70174    +49 0711 2842222                None  leonekohler@surfeu.de   
2    H2G 1A7   +1 (514) 721-4711                None    ftremblay@gmail.com   

   SupportRepId  
0             3  
1             5  
2             3

In [4]:
pd.read_sql(
    """
    SELECT *
    FROM Invoice
    LIMIT 3
    """, conn
)

InvoiceId  CustomerId          InvoiceDate           BillingAddress  \
0          1           2  2009-01-01 00:00:00  Theodor-Heuss-Straße 34   
1          2           4  2009-01-02 00:00:00         Ullevålsveien 14   
2          3           8  2009-01-03 00:00:00          Grétrystraat 63   

  BillingCity BillingState BillingCountry BillingPostalCode  Total  
0   Stuttgart         None        Germany             70174   1.98  
1        Oslo         None         Norway              0171   3.96  
2    Brussels         None        Belgium              1000   5.94

### Task 1: Write a query to return the customer IDs for every customer in Germany.

In [6]:
#

pd.read_sql(
    """
    SELECT CustomerId
    FROM Customer 
    WHERE Country = 'Germany'
    """, conn
)

CustomerId
0           2
1          36
2          37
3          38

### Task 2: Write a query to return the last names, email addresses and *billing* addresses of the customers who live in Canada.

In [7]:
# 

pd.read_sql(
    """
    SELECT c.LastName, c.Email, i.BillingAddress
    FROM Customer c
    INNER JOIN Invoice i 
    ON c.CustomerId = i.CustomerId
    WHERE c.Country = 'Canada'
    """, conn
)

LastName                   Email          BillingAddress
0   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
1   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
2   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
3   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
4   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
5   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
6   Tremblay     ftremblay@gmail.com       1498 rue Bélanger
7    Philips      mphilips12@shaw.ca          8210 111 ST NW
8    Philips      mphilips12@shaw.ca          8210 111 ST NW
9    Philips      mphilips12@shaw.ca          8210 111 ST NW
10   Philips      mphilips12@shaw.ca          8210 111 ST NW
11   Philips      mphilips12@shaw.ca          8210 111 ST NW
12   Philips      mphilips12@shaw.ca          8210 111 ST NW
13   Philips      mphilips12@shaw.ca          8210 111 ST NW
14  Peterson     jenniferp@rogers.ca     700 W Pender Street
15  Peterson     jenniferp@rogers.ca     700 W Pender Street
16  Peterson     jenniferp@rogers.ca     700 W Pender Street
17  Peterson     jenniferp@rogers.ca     700 W Pender Street
18  Peterson     jenniferp@rogers.ca     700 W Pender Street
19  Peterson     jenniferp@rogers.ca     700 W Pender Street
20  Peterson     jenniferp@rogers.ca     700 W Pender Street
21     Brown        robbrown@shaw.ca  796 Dundas Street West
22     Brown        robbrown@shaw.ca  796 Dundas Street West
23     Brown        robbrown@shaw.ca  796 Dundas Street West
24     Brown        robbrown@shaw.ca  796 Dundas Street West
25     Brown        robbrown@shaw.ca  796 Dundas Street West
26     Brown        robbrown@shaw.ca  796 Dundas Street West
27     Brown        robbrown@shaw.ca  796 Dundas Street West
28   Francis     edfrancis@yachoo.ca        230 Elgin Street
29   Francis     edfrancis@yachoo.ca        230 Elgin Street
30   Francis     edfrancis@yachoo.ca        230 Elgin Street
31   Francis     edfrancis@yachoo.ca        230 Elgin Street
32   Francis     edfrancis@yachoo.ca        230 Elgin Street
33   Francis     edfrancis@yachoo.ca        230 Elgin Street
34   Francis     edfrancis@yachoo.ca        230 Elgin Street
35      Silk    marthasilk@gmail.com   194A Chain Lake Drive
36      Silk    marthasilk@gmail.com   194A Chain Lake Drive
37      Silk    marthasilk@gmail.com   194A Chain Lake Drive
38      Silk    marthasilk@gmail.com   194A Chain Lake Drive
39      Silk    marthasilk@gmail.com   194A Chain Lake Drive
40      Silk    marthasilk@gmail.com   194A Chain Lake Drive
41      Silk    marthasilk@gmail.com   194A Chain Lake Drive
42  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
43  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
44  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
45  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
46  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
47  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
48  Mitchell  aaronmitchell@yahoo.ca      696 Osborne Street
49  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
50  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
51  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
52  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
53  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
54  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street
55  Sullivan  ellie.sullivan@shaw.ca          5112 48 Street

### Task 3: Write a query to return the billing totals by country, in descending order of the totals.

In [12]:
## remember that the totals per row aka per line are for an individual customer but 
## we want the total per country which means when several customers are from the same 
## country we have to sum them up, hence why we use SUM below

In [11]:
 #
pd.read_sql(
    """
    SELECT BillingCountry, SUM(Total) AS sum
    FROM Invoice
    GROUP BY BillingCountry
    ORDER BY sum DESC 
    """, conn
)

BillingCountry     sum
0              USA  523.06
1           Canada  303.96
2           France  195.10
3           Brazil  190.10
4          Germany  156.48
5   United Kingdom  112.86
6   Czech Republic   90.24
7         Portugal   77.24
8            India   75.26
9            Chile   46.62
10         Ireland   45.62
11         Hungary   45.62
12         Austria   42.62
13         Finland   41.62
14     Netherlands   40.62
15          Norway   39.62
16          Sweden   38.62
17          Poland   37.62
18           Italy   37.62
19         Denmark   37.62
20       Australia   37.62
21       Argentina   37.62
22           Spain   37.62
23         Belgium   37.62

### ***Task 4: Write a query to return the last names, email addresses, billing addresses, and most recent invoice dates for the customers whose most recent invoice is after Dec 31, 2011.

What the solution notebook gave >>

pd.read_sql(
    """
    WITH t AS
    (SELECT CustomerId, BillingAddress, MAX(InvoiceDate) as latestInvoice
    FROM Invoice
    WHERE InvoiceDate > "2011-12-31"
    GROUP BY CustomerId)
    
    SELECT LastName, Email, latestInvoice
    FROM Customer, t
    GROUP BY Customer.CustomerId
    """, conn
)

BUT I understand what i wrote below

In [19]:
# 

pd.read_sql(
    """
    SELECT c.LastName, c.Email, i.BillingAddress, i.InvoiceDate
    FROM Invoice i
    INNER JOIN Customer c 
    ON i.CustomerId = c.CustomerId  
    WHERE i.InvoiceDate > "2011-12-31"
    """, conn
)

LastName                      Email  \
0        Taylor       mark.taylor@yahoo.au   
1       Martins   eduardo@woodstock.com.br   
2         Rocha           alero@uol.com.br   
3         Ramos   fernadaramos4@uol.com.br   
4      Peterson        jenniferp@rogers.ca   
..          ...                        ...   
158     Stevens         vstevens@yahoo.com   
159       Brown           robbrown@shaw.ca   
160     Sampaio          masampaio@sapo.pt   
161  Hämäläinen  terhi.hamalainen@apple.fi   
162      Pareek    manoj.pareek@rediff.com   

                               BillingAddress          InvoiceDate  
0                           421 Bourke Street  2012-01-01 00:00:00  
1                   Rua Dr. Falcão Filho, 155  2012-01-09 00:00:00  
2                          Av. Paulista, 2022  2012-01-22 00:00:00  
3                                Qe 7 Bloco G  2012-01-22 00:00:00  
4                         700 W Pender Street  2012-01-23 00:00:00  
..                                        ...                  ...  
158                     319 N. Frances Street  2013-12-05 00:00:00  
159                    796 Dundas Street West  2013-12-06 00:00:00  
160  Rua dos Campeões Europeus de Viena, 4350  2013-12-09 00:00:00  
161                           Porthaninkatu 9  2013-12-14 00:00:00  
162                       12,Community Centre  2013-12-22 00:00:00  

[163 rows x 4 columns]

## Part 2: Bayesian Statistics

$$ P(A|B) = \dfrac{P(B|A)P(A)}{P(B)}$$

### The Prompt: 
> In the second session of the 98th congress in 1984, a congressman voted 'No' on providing aid to El Salvador. 
> 
> Given that 60% of the congress were Democrats, 75% of whom voted 'No' for providing aid to El Salvador, and only 5% of Republicans voted 'No' to the proposal, what is the conditional probability that this individual is a Democrat?

### Task 5: Calculate the probability

In [20]:
# 
p_dem = .6
p_rep = 1-p_dem

p_no_given_dem = .75
p_no_given_rep = .05

p_no = p_no_given_dem * p_dem + p_no_given_rep * p_rep

p_dem_given_no = p_no_given_dem * p_dem / p_no
p_dem_given_no

0.9574468085106382

There is a 96% conditional probability that this individual is a Democrat

**1. Which probability are we trying to find?**

 - P(Democrat | No)   
   

**2. Based on that, what other pieces do we need?**

   - P(Democrat | No) = P(No | Democrat) * P(Democrat) / P(No)
   - SO:
   - P(No | Democrat) = .75
   - P(Democrat) = .6
   - P(No) = P(No | Dem) * P(Dem) + P(No | Rep) * P(Rep) = (.75 * .6) + (.06 * (1 - .6)) = .47

**3. Result?**

   - P(Democrat | No) = P(No | Democrat) * P(Democrat) / P(No)
   - P(Democrat | No) = .75 * .6 / .47 = .9574 &#8594; **96%!**

## Part 3: Normal Distributions

In this part, you will analyze hours of sleep for Freshman at a University. We know that the population distribution of hours of sleep for Freshman at the University is normally distributed with a mean of 6 and a standard deviation of 2.

### Task 6: Create a numeric variable `z_eight` that contains the z-score for a Freshman that sleeps 8 hours 

HERE we use formula for single point z score:

z = x - mu/ sigma

- x = single point = 8
- mu = population mean = 6
- sigma = population std deviation = 2





Formula for sample mean z score:

z = x_bar - mu/ s

- when we see x_bar we know to use sample mean instead of pop mean for mu
- s = sample std deviation

In [21]:
# Code here

mu_fresh = 6
std_fresh = 2

# z_score = point - mu / std

z_eight = (8 - mu_fresh)/std_fresh
z_eight

1.0

### Task 7: Create a numeric variable `over_eight` containing the approximate proportion of Freshman who get more than 8 hours of sleep

1. Step 1, import package
- from scipy import stats
2. Step 2, write code formula
 - choose one from below

default is:
- stats.norm.cdf(z_score)
- used when we look at left side aka less than aka decreasing


HERE we want:

- 1 - stats.norm.cdf(z_score)
- used when we look at right side aka more than aka increasing

In [25]:
# Code here

from scipy import stats

over_eight = 1 - stats.norm.cdf(z_eight)
over_eight

0.15865525393145707

### Task 8: Calculate a 95% confidence interval for the amount of sleep for one Freshman who sleeps 7 times with a mean of 5 hours and a standard deviation of 0.5 hours

In [27]:
# Code here

n = 7
mean = 5
std = 0.5

# Still need to calculate that standard error by hand
sterr = std / (n ** 0.5)

# Using Scipy Stats
# Using t because n=7

stats.t.interval(alpha=.95,  # Confidence level
                 loc=mean,  # Mean
                 scale=sterr,
                 df=n-1)

(4.53757712628571, 5.46242287371429)

### Task 9: Interpret the 95% Confidence Interval that you derived in the above question

Interpret: 

- A 95% confidence interval means that there is a 95% chance for the interval to contain the true population mean. In this case, that means there is a 95% chance that the true mean of the distribution of hours of sleep likely for this Freshman is between 4.9 and 5.04 hours.


## Part 4: Statistical Testing

Let's make some hypotheses about cars!

### Task 10: Set Up

In [ ]:
# Imports - need pandas and matplotlib


In [ ]:
# Import the auto-mpg csv from the data folder
df = None

In [ ]:
# Check the head of our data


In [ ]:
# Check the info on our data


In [ ]:
# Describe the data


In [ ]:
# Run this cell without changes to set up for the next set of questions

# Grabbing the median car weight
weight_median = df['weight'].median()

# Separating based on light cars and heavy cars
heavy_cars = df.loc[df['weight'] > weight_median].sample(n=100, random_state=42)
light_cars = df.loc[df['weight'] <= weight_median].sample(n=100, random_state=42)

# Plotting histograms for acceleration for both light and heavy cars
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))

ax1.set_title('Sample of Heavy Cars')
ax1.set_xlabel('Acceleration')
ax1.set_ylabel('Frequency')
ax1.hist(heavy_cars['acceleration'], bins=12)

ax2.set_title('Sample of Light Cars')
ax2.set_xlabel('Acceleration')
ax2.set_ylabel('Frequency')
ax2.hist(light_cars['acceleration'], bins=12)
plt.show()

### The Prompt: 

> Car manufacturers want to know if light cars accelerate faster than heavy cars.

### Task 11: Set up the null and alternative hypotheses for this test

**Null hypothesis:**

- 


**Alternative hypothesis:**

- 


### Task 12: Describe your Type 1 and Type 2 errors in this context

**Type I Error:**

- 


**Type II Error:**

- 



### Task 13: The Hypothesis Test!

### 13a: Describe the type of test:

Description:

- 


### 13b: Run the test, using alpha = .05 and assuming that the samples have equal variance

In [ ]:
# Code here


### 13c: Analyze your results: can you reject the null hypothesis?

Can you reject the null hypothesis?

- 
